File: containerWithMostWater.cpp
// 11 Array medium

#include <vector>
class Solution {
  public:
    int maxArea( vector<int> &height ) {
        int i = 0, j = height.size() - 1;
        int m = 0;
        while ( j > i ) {
            m = max( m, min( height[ i ], height[ j ] ) * ( j - i ) );
            if ( height[ i ] < height[ j ] ) {
                i++;
            } else {
                j--;
            }
        }
        return m;
    }
};

File: containerWithMostWater.js
// 11 greedy medium

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
  let left = 0, right = height.length-1
  let ret = 0
  while (left < right) {
    const sum = (right-left) * Math.min(height[left], height[right])
    if (sum > ret) ret = sum
    if (height[left] > height[right]) right--
    else left++
  }
  return ret
}

console.log(maxArea([1,8,6,2,5,4,8,3,7]))
console.log(maxArea([1,2,4,3]))